File: design-patterns/java/templatemethod/ExcelReportExporter.java
import java.util.Map;

public class ExcelReportExporter extends AbstractReportExporter {
    //prepareData() not overridden - default will be used
    //openFile() not overridden - default will be used

    @Override
    protected void writeHeader(ReportData data) {
        System.out.println("Excel: Writing header: " + String.join(",", data.getHeaders()));
    }

    @Override
    protected void writeDataRows(ReportData data) {
        System.out.println("Excel: Writing data rows...");
        for (Map<String, Object> row : data.getRows()) {
            System.out.println("Excel: " + row.values());
        }
    }

    // writeFooter() not overridden - default will be used
    // closeFile() not overridden - default will be used
}

File: design-patterns/cpp/templatemethod/excel_report_exporter.cpp
#include "excel_report_exporter.h"
#include <iostream>

void ExcelReportExporter::writeHeader(const ReportData& data) {
    std::cout << "Writing Excel header..." << std::endl;
    std::cout << "| ";
    for (const auto& header : data.getHeaders()) {
        std::cout << header << " | ";
    }
    std::cout << std::endl;
    std::cout << "|";
    for (size_t i = 0; i < data.getHeaders().size(); ++i) {
        std::cout << "---|";
    }
    std::cout << std::endl;
}

void ExcelReportExporter::writeDataRows(const ReportData& data) {
    std::cout << "Writing Excel data rows..." << std::endl;
    for (const auto& row : data.getRows()) {
        std::cout << "| ";
        for (const auto& header : data.getHeaders()) {
            const auto& value = row.at(header);
            std::visit([](const auto& v) { std::cout << v; }, value);
            std::cout << " | ";
        }
        std::cout << std::endl;
    }
} 